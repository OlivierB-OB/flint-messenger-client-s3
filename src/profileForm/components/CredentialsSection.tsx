import FormControl from '@material-ui/core/FormControl';
import FormGroup from '@material-ui/core/FormGroup';
import FormLabel from '@material-ui/core/FormLabel';
import TextField from '@material-ui/core/TextField';
import React from 'react';
import { PasswordRequirements } from './PasswordRequirements';
import { IFormField, IPasswordField, IProfileFormFields } from '../types';

export interface ICredentialsSectionProps {
  password: IPasswordField;
  confirmation: IFormField<string>;
  update<T extends keyof IProfileFormFields>(field: T, value: IProfileFormFields[T]['value']): void;
}

export function CredentialsSection({ password, confirmation, update }: ICredentialsSectionProps) {
  return (
    <FormControl component="fieldset" fullWidth={true}>
      <FormLabel component="legend" style={{ margin: '1rem 0' }}>
        Your credentials:
      </FormLabel>
      <FormGroup>
        <TextField
          type="password"
          label="Password"
          value={password.value}
          fullWidth={true}
          onChange={(event) => update('password', event.target.value)}
          {...(!password.isValid ? { error: true, helperText: password.error } : {})}
        />
        <TextField
          type="password"
          label="Password confirmation"
          value={confirmation.value}
          required={!!password.value}
          fullWidth={true}
          onChange={(event) => update('confirmation', event.target.value)}
          {...(!confirmation.isValid ? { error: true, helperText: confirmation.error } : {})}
        />
        <PasswordRequirements password={password} />
      </FormGroup>
    </FormControl>
  );
}
