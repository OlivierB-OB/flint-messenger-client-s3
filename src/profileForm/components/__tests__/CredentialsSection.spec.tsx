import TextField from '@material-ui/core/TextField';
import React from 'react';
import { shallow } from 'enzyme';
import { CredentialsSection } from '../CredentialsSection';

describe('CredentialsSection', () => {
  it('should display the password and confiramtion value', async () => {
    const data: any = {
      password: {
        value: 'password value',
        isValid: true,
      },
      confirmation: {
        value: 'confirmation value',
        isValid: true,
      },
    };
    const component = shallow(<CredentialsSection {...data} />);
    expect(component).toMatchSnapshot();
  });

  it('should show error if any', async () => {
    const data: any = {
      password: {
        value: 'password value',
        isValid: false,
        error: 'password error',
      },
      confirmation: {
        value: 'confirmation value',
        isValid: false,
        error: 'confirmation error',
      },
    };
    const component = shallow(<CredentialsSection {...data} />);
    expect(component).toMatchSnapshot();
  });

  it('should display the confiramtion as optional if the password field is empty', async () => {
    const data: any = {
      password: {
        value: '',
        isValid: true,
      },
      confirmation: {
        value: '',
        isValid: true,
      },
    };
    const component = shallow(<CredentialsSection {...data} />);
    expect(component).toMatchSnapshot();
  });

  it('should allow editing the password value', async () => {
    const data: any = {
      password: {
        value: 'password value',
        isValid: true,
      },
      confirmation: {
        value: 'confirmation value',
        isValid: true,
      },
      update: jest.fn(),
    };
    const component = shallow(<CredentialsSection {...data} />);
    component
      .find(TextField)
      .at(0)
      .simulate('change', { target: { value: 'new value' } });
    expect(data.update).toHaveBeenCalledWith('password', 'new value');
  });

  it('should allow editing the confirmation value', async () => {
    const data: any = {
      password: {
        value: 'password value',
        isValid: true,
      },
      confirmation: {
        value: 'confirmation value',
        isValid: true,
      },
      update: jest.fn(),
    };
    const component = shallow(<CredentialsSection {...data} />);
    component
      .find(TextField)
      .at(1)
      .simulate('change', { target: { value: 'new value' } });
    expect(data.update).toHaveBeenCalledWith('confirmation', 'new value');
  });
});
