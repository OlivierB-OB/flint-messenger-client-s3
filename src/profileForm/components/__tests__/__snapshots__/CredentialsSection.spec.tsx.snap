// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CredentialsSection should display the confiramtion as optional if the password field is empty 1`] = `
<WithStyles(ForwardRef(FormControl))
  component="fieldset"
  fullWidth={true}
>
  <WithStyles(ForwardRef(FormLabel))
    component="legend"
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    Your credentials:
  </WithStyles(ForwardRef(FormLabel))>
  <WithStyles(ForwardRef(FormGroup))>
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Password"
      onChange={[Function]}
      type="password"
      value=""
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Password confirmation"
      onChange={[Function]}
      required={false}
      type="password"
      value=""
    />
    <PasswordRequirements
      password={
        Object {
          "isValid": true,
          "value": "",
        }
      }
    />
  </WithStyles(ForwardRef(FormGroup))>
</WithStyles(ForwardRef(FormControl))>
`;

exports[`CredentialsSection should display the password and confiramtion value 1`] = `
<WithStyles(ForwardRef(FormControl))
  component="fieldset"
  fullWidth={true}
>
  <WithStyles(ForwardRef(FormLabel))
    component="legend"
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    Your credentials:
  </WithStyles(ForwardRef(FormLabel))>
  <WithStyles(ForwardRef(FormGroup))>
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Password"
      onChange={[Function]}
      type="password"
      value="password value"
    />
    <WithStyles(ForwardRef(TextField))
      fullWidth={true}
      label="Password confirmation"
      onChange={[Function]}
      required={true}
      type="password"
      value="confirmation value"
    />
    <PasswordRequirements
      password={
        Object {
          "isValid": true,
          "value": "password value",
        }
      }
    />
  </WithStyles(ForwardRef(FormGroup))>
</WithStyles(ForwardRef(FormControl))>
`;

exports[`CredentialsSection should show error if any 1`] = `
<WithStyles(ForwardRef(FormControl))
  component="fieldset"
  fullWidth={true}
>
  <WithStyles(ForwardRef(FormLabel))
    component="legend"
    style={
      Object {
        "margin": "1rem 0",
      }
    }
  >
    Your credentials:
  </WithStyles(ForwardRef(FormLabel))>
  <WithStyles(ForwardRef(FormGroup))>
    <WithStyles(ForwardRef(TextField))
      error={true}
      fullWidth={true}
      helperText="password error"
      label="Password"
      onChange={[Function]}
      type="password"
      value="password value"
    />
    <WithStyles(ForwardRef(TextField))
      error={true}
      fullWidth={true}
      helperText="confirmation error"
      label="Password confirmation"
      onChange={[Function]}
      required={true}
      type="password"
      value="confirmation value"
    />
    <PasswordRequirements
      password={
        Object {
          "error": "password error",
          "isValid": false,
          "value": "password value",
        }
      }
    />
  </WithStyles(ForwardRef(FormGroup))>
</WithStyles(ForwardRef(FormControl))>
`;
